File: algorithms/2_medium/P/00086_partition_list/solution.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* partition(ListNode* head, int x) {
        ListNode dummy1 = ListNode(0);
        ListNode dummy2 = ListNode(0);
        ListNode* prev1 = &dummy1;
        ListNode* prev2 = &dummy2;
        ListNode* curr = head;

        while (curr != nullptr) {
            ListNode* next = curr->next;
            curr->next = nullptr;

            if (curr->val < x) {
                prev1->next= curr;
                prev1 = curr;
            } else {
                prev2->next= curr;
                prev2 = curr;
            }

            curr = next;
        }

        prev1->next = dummy2.next;

        return dummy1.next;
    }
};

File: algorithms/2_medium/P/00086_partition_list/solution.rs
// Definition for singly-linked list.
// #[derive(PartialEq, Eq, Clone, Debug)]
// pub struct ListNode {
//   pub val: i32,
//   pub next: Option<Box<ListNode>>
// }
// 
// impl ListNode {
//   #[inline]
//   fn new(val: i32) -> Self {
//     ListNode {
//       next: None,
//       val
//     }
//   }
// }
impl Solution {
    pub fn partition(head: Option<Box<ListNode>>, x: i32) -> Option<Box<ListNode>> {
        let mut dummy1 = ListNode::new(0);
        let mut dummy2 = ListNode::new(0);
        let mut prev1 = &mut dummy1;
        let mut prev2 = &mut dummy2;
        let mut curr_opt = head;

        while let Some(mut curr) = curr_opt {
            let next = curr.next.take();

            if curr.val < x {
                prev1.next = Some(curr);
                prev1 = prev1.next.as_mut().unwrap().as_mut();
            } else {
                prev2.next = Some(curr);
                prev2 = prev2.next.as_mut().unwrap().as_mut();
            }

            curr_opt = next;
        }

        prev1.next = dummy2.next;

        dummy1.next
    }
}
